### DISCUSSION TITLE
Including unharmonized data in results

### DISCUSSION THREAD

**Discussion by hbeale** [CBIIT/ccdi-federation-api#21] [API Design Discussion]
@e-t-k  and I have been talking about what it might look like to include unharmonized data in result. Our assessment is that a user with the json can relatively easily go from harmonized -> unharmonized because there is an 'ancestor' field. If you disagree with our assessment, please let us know. 

We also think it's valuable for the NCI front end team to display related unharmonized values when a user is searching for a harmonized one.

It might be harder to go from unharmonized data to related harmonized data, but I can't think of a use case for that.

**Comment by e-t-k** [comment]
Thank you Holly.
In addition it's important that the front end provide filtering / searching directly on the unharmonized fields. 

It may be unreasonable to list every unharmonized field in the front end GUI by default, especially since new ones can appear ad hoc with the release of new Samples/Subjects. However one option could be to, when filtering on a value of a harmonized field, in addition filter on any unharmonized values that appear in the ancestors entry for that field.

For example, I understand there was consideration at the data harmonization meeting of defining harmonized fields to allow only a subset of a CDE's permissible values and thus avoid confusion about which to select when multiple may be applicable (see [Clay's examples in Slack](https://uc-cri.slack.com/archives/C05FFPKLQ79/p1698254766364549).) 

If implemented, we might have a harmonized field eg [CDE 6347743](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=6347743%20and%20ver_nr=1) which permits only more general values eg "PolyA" and "rRNA Depletion".

However, unharmonized fields which appear as ancestors of that field can include the more specific "Oligo-dT" or "Poly-T Enrichment" values. Once the user filters the harmonized field to "PolyA" (or even before), those values should be made available for further filtering.

**Ancestors key**
It's notable that this key, currently, can vary on a per-Sample/Subject basis.
Subject 1 `race.ancestors` can include `unharmonized.attr_race`, while the same server's Subject 2's `race.ancestors` includes `unharmonized.Race` instead.
This concerned me at first but actually it shouldn't cause implementation difficulties, since it's already expected to differ between servers. And it will be useful for Treehouse or any site sharing samples from various third parties with inconsistent field names.

**Comment by claymcleod** [comment]
Everything you said here makes sense to me.

With respect to matching an unharmonized key back to all harmonized keys it influences, my thought was that those relationships can always be calculated on the fly. Adding the link from both directions will introduce some amount of bloat in the response simply to store redundant data. In my mind, the process of aggregating the data in the frontend would have to be quite computationally complex before it made sense to add that overhead to every response.

I know you're not advocating for that here, but just thinking it through out loud.

**Comment by claymcleod** [comment]
We discussed this during our meeting on 11/03/2023, and it seemed that everyone agreed with the points for future implementers to consider. As for the issue of an unharmonized field points to all of the harmonized fields it influences, it seems the group was okay with the logic I set out above (please comment if you disagree). In the meantime, I'm going to close this discussion unless there are further points to consider.
